algebird-core/src/main/scala/com/twitter/algebird/experimental/BloomFilter.scala

```diff
@@ -0,0 +1,413 @@
+  }
+
+  /**
+   * Cardinality estimates are taken from Theorem 1 on page 15 of
+   * "Cardinality estimation and dynamic length adaptation for Bloom filters"
+   * by Papapetrou, Siberski, and Nejdl:
+   * http://www.softnet.tuc.gr/~papapetrou/publications/Bloomfilters-DAPD.pdf
+   *
+   * Roughly, by using bounds on the expected number of true bits after n elements
+   * have been inserted into the Bloom filter, we can go from the actual number of
+   * true bits (which is known) to an estimate of the cardinality.
+   *
+   * approximationWidth defines an interval around the maximum-likelihood cardinality
+   * estimate. Namely, the approximation returned is of the form
+   * (min, estimate, max) =
+   *   ((1 - approxWidth) * estimate, estimate, (1 + approxWidth) * estimate)
+   */
+  def sizeEstimate(
+      numBits: Int,
+      numHashes: Int,
+      width: Int,
+      approximationWidth: Double = 0.05
+  ): Approximate[Long] = {
+    assert(0 <= approximationWidth && approximationWidth < 1, "approximationWidth must lie in [0, 1)")
+
+    /**
+     * s(n) is the expected number of bits that have been set to true after
+     * n elements have been inserted into the Bloom filter.
+     * This is \hat{S}(n) in the cardinality estimation paper used above.
+     */
+    def s(n: Int): Double =
+      width * (1 - scala.math.pow(1 - 1.0 / width, numHashes * n))
+
+    /**
+     * sInverse(t) is the maximum likelihood value for the number of elements
+     * that have been inserted into the Bloom filter when it has t bits set to true.
+     * This is \hat{S}^{-1}(t) in the cardinality estimation paper used above.
+     */
+    def sInverse(t: Int): Double =
+      scala.math.log1p(-t.toDouble / width) / (numHashes * scala.math.log1p(-1.0 / width))
+
+    // Variable names correspond to those used in the paper.
+    val t = numBits
+    val n = sInverse(t).round.toInt
+    // Take the min and max because the probability formula assumes
+    // nl <= sInverse(t - 1) and sInverse(t + 1) <= nr
+    val nl =
+      scala.math.min(sInverse(t - 1).floor, (1 - approximationWidth) * n).toInt
+    val nr =
+      scala.math.max(sInverse(t + 1).ceil, (1 + approximationWidth) * n).toInt
+    val prob =
+      1 -
+        scala.math.exp(t - 1 - s(nl)) *
+          scala.math.pow(s(nl) / (t - 1), t - 1) -
+        scala.math.exp(-scala.math.pow(t + 1 - s(nr), 2) / (2 * s(nr)))
+
+    Approximate[Long](nl, n, nr, scala.math.max(0, prob))
+  }
+}
+
+/**
+ * Bloom Filter - a probabilistic data structure to test presence of an element.
+ *
+ * Operations
+ *   1) insert: hash the value k times, updating the bitfield at the index equal to each hashed value
+ *   2) query: hash the value k times.  If there are k collisions, then return true; otherwise false.
+ *
+ * http://en.wikipedia.org/wiki/Bloom_filter
+ */
+final case class BloomFilter[A](n: Int, w: Int)(implicit val hash: Hash128[A]) { self =>
+  def hashToArray(s: A): Array[Int] = {
```

Comment:
lets add a comment here as to what is going on: what is the size of the result, etc...

Reply:
also, maybe worth it to consider a private variant that does zero allocation so we can use that one in sumOption. I think that will be a significant win.